src/praisonai-agents/praisonaiagents/agent/chat_mixin.py

```
@@ -29,6 +29,129 @@
+                return str(value).strip()
+        return ""
+
+    def _format_memory_prefetch(self, results: Any, limit: int, token_budget: int) -> str:
+        """Format deduplicated memory hits inside a strict estimated-token budget."""
+        if not isinstance(results, (list, tuple)):
+            return ""
+
+        try:
+            from ..context.tokens import estimate_tokens_heuristic
+        except ImportError:  # pragma: no cover - core module is always present
+            estimate_tokens_heuristic = lambda text: max(1, len(text) // 4)
```

Ruff (0.16.1)

[error] Do not assign a `lambda` expression, use a `def`

Rewrite `estimate_tokens_heuristic` as a `def`

(E731)